README.md
<p align="center">
  <a href="https://github.com/wellyshen/react-cool-virtual/blob/master/README.md" title="React Cool Virtual"><img src="https://github.com/wellyshen/react-cool-virtual/blob/master/images/logo.svg" alt="React Cool Virtual"></a>
</p>

<p align="center">A tiny React hook for rendering large datasets like a breeze.</p>

<div align="center">

[![npm version](https://img.shields.io/npm/v/react-cool-virtual?style=flat-square)](https://www.npmjs.com/package/react-cool-virtual)
[![npm downloads](https://img.shields.io/npm/dt/react-cool-virtual?style=flat-square)](https://www.npmtrends.com/react-cool-virtual)
[![coverage status](https://img.shields.io/coveralls/github/wellyshen/react-cool-virtual?style=flat-square)](https://coveralls.io/github/wellyshen/react-cool-virtual?branch=master)
[![gzip size](https://badgen.net/bundlephobia/minzip/react-cool-virtual?label=gzip%20size&style=flat-square)](https://bundlephobia.com/result?p=react-cool-virtual)
[![best of js](https://img.shields.io/endpoint?style=flat-square&url=https://bestofjs-serverless.now.sh/api/project-badge?fullName=wellyshen%2Freact-cool-virtual)](https://bestofjs.org/projects/react-cool-virtual)
[![All Contributors](https://img.shields.io/badge/all_contributors-4-orange.svg?style=flat-square)](#contributors-)

</div>

## Features <!-- omit in toc -->

- ♻️ Renders millions of items with highly performant way, using [DOM recycling](https://developers.google.com/web/updates/2016/07/infinite-scroller).
- 🎣 Easy to use, based on React [hook](https://reactjs.org/docs/hooks-custom.html#using-a-custom-hook).
- 💅🏼 Apply styles without hassle, just [few setups](#basic-usage).
- 🧱 Supports [fixed](#fixed-size), [variable](#variable-size), [dynamic](#dynamic-size), and [real-time](#real-time-resize) heights/widths.
- 🖥 Supports [responsive web design (RWD)](#responsive-web-design-rwd) for better UX.
- 📌 Supports [sticky headers](#sticky-headers) for building on-trend lists.
- 🚚 Built-ins [load more callback](#infinite-scroll) for you to deal with infinite scroll + [skeleton screens](https://uxdesign.cc/what-you-should-know-about-skeleton-screens-a820c45a571a).
- 🖱 Imperative [scroll-to methods](#scroll-to-offset--items) for offset, items, and alignment.
- 🛹 Out-of-the-box [smooth scrolling](#smooth-scrolling) and the effect is DIY-able.
- 💬 It's possible to implement [stick to bottom](#sticking-to-bottom) and [pre-pending items](#pre-pending-items) for chat, feeds, etc.
- ⛳ Provides `isScrolling` indicator to you for UI placeholders or [performance optimization](#use-isscrolling-indicator).
- 🗄️ Supports [server-side rendering (SSR)](#server-side-rendering-ssr) for a fast [FP + FCP](https://developers.google.com/web/updates/2019/02/rendering-on-the-web#server-rendering) and better [SEO](https://developers.google.com/web/updates/2019/02/rendering-on-the-web#server-rendering).
- 📜 Supports [TypeScript](#working-in-typescript) type definition.
- 🎛 Super flexible [API](#api) design, built with DX in mind.
- 🦔 Tiny size ([~ 3.1kB gzipped](https://bundlephobia.com/result?p=react-cool-virtual)). No external dependencies, aside for the `react`.

## Why? <!-- omit in toc -->

When rendering a large set of data (e.g. list, table, etc.) in React, we all face performance/memory troubles. There're [some great libraries](https://www.npmjs.com/search?q=react%20virtualized) already available but most of them are component-based solutions that provide well-defineded way of using but increase a lot of bundle size. However, [a library](https://github.com/tannerlinsley/react-virtual) comes out as a hook-based solution that is flexible and `headless` but using and styling it can be verbose (because it's a low-level hook). Furthermore, it lacks many of the [useful features](#features).

React Cool Virtual is a [tiny](https://bundlephobia.com/result?p=react-cool-virtual) React hook that gives you a **better DX** and **modern way** for virtualizing a large amount of data without struggle 🤯.

## Docs <!-- omit in toc -->

- [Getting Started](#getting-started)
  - [Requirement](#requirement)
  - [Installation](#installation)
  - [CDN](#cdn)
  - [Basic Usage](#basic-usage)
- [Examples](#examples)
  - [Fixed Size](#fixed-size)
  - [Variable Size](#variable-size)
  - [Dynamic Size](#dynamic-size)
  - [Real-time Resize](#real-time-resize)
  - [Responsive Web Design (RWD)](#responsive-web-design-rwd)
  - [Sticky Headers](#sticky-headers)
  - [Scroll to Offset / Items](#scroll-to-offset--items)
  - [Smooth Scrolling](#smooth-scrolling)
  - [Infinite Scroll](#infinite-scroll)
  - [Pre-pending Items](#pre-pending-items)
  - [Filtering Items](#filtering-items)
  - [Sticking to Bottom](#sticking-to-bottom)
  - [Working with Input Elements](#working-with-input-elements)
  - [Dealing with Dynamic Items](#dealing-with-dynamic-items)
  - [Server-side Rendering (SSR)](#server-side-rendering-ssr)
- [API](#api)
  - [Options](#options)
  - [Return Values](#return-values)
- [Others](#others)
  - [Performance Optimization](#performance-optimization)
  - [How to Share A `ref`?](#how-to-share-a-ref)
  - [Layout Items](#layout-items)
  - [Working in TypeScript](#working-in-typescript)
  - [ResizeObserver Polyfill](#resizeobserver-polyfill)

## Getting Started

### Requirement

To use React Cool Virtual, you must use `react@16.8.0` or greater which includes hooks.

### Installation

This package is distributed via [npm](https://www.npmjs.com/package/react-cool-virtual).

```sh
$ yarn add react-cool-virtual
# or
$ npm install --save react-cool-virtual
```

> ⚠️ This package using [ResizeObserver API](https://developer.mozilla.org/en-US/docs/Web/API/Resize_Observer_API) under the hook. [Most modern browsers support it natively](https://caniuse.com/?search=ResizeObserver), you can also add [polyfill](#resizeobserver-polyfill) for full browser support.

### CDN

If you're not using a module bundler or package manager. We also provide a [UMD](https://github.com/umdjs/umd) build which is available over the [unpkg.com](https://unpkg.com) CDN. Simply use a `<script>` tag to add it after [React CDN links](https://reactjs.org/docs/cdn-links.html) as below:

<!-- prettier-ignore-start -->
```html
<script crossorigin src="https://unpkg.com/react/umd/react.production.min.js"></script>
<script crossorigin src="https://unpkg.com/react-dom/umd/react-dom.production.min.js"></script>
<!-- react-cool-virtual comes here -->
<script crossorigin src="https://unpkg.com/react-cool-virtual/dist/index.umd.production.min.js"></script>
```
<!-- prettier-ignore-end -->

Once you've added this you will have access to the `window.ReactCoolVirtual.useVirtual` variable.

### Basic Usage

Here's the basic concept of how it rocks:

```js
import useVirtual from "react-cool-virtual";

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 10000, // Provide the total number for the list items
    itemSize: 50, // The size of each item (default = 50)
  });

  return (
    <div
      ref={outerRef} // Attach the `outerRef` to the scroll container
      style={{ width: "300px", height: "500px", overflow: "auto" }}
    >
      {/* Attach the `innerRef` to the wrapper of the items */}
      <div ref={innerRef}>
        {items.map(({ index, size }) => (
          // You can set the item's height with the `size` property
          <div key={index} style={{ height: `${size}px` }}>
            ⭐️ {index}
          </div>
        ))}
      </div>
    </div>
  );
};
```

✨ Pretty easy right? React Cool Virtual is more powerful than you think. Let's explore more use cases through the examples!

## Examples

### Fixed Size

This example demonstrates how to create a fixed size row. For column or grid, please refer to CodeSandbox.

[![Edit RCV - Fixed Size](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-fixed-size-bowcu?fontsize=14&hidenavigation=1&theme=dark)

```js
import useVirtual from "react-cool-virtual";

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 1000,
  });

  return (
    <div
      style={{ width: "300px", height: "300px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, size }) => (
          <div key={index} style={{ height: `${size}px` }}>
            ⭐️ {index}
          </div>
        ))}
      </div>
    </div>
  );
};
```

### Variable Size

This example demonstrates how to create a variable size row. For column or grid, please refer to CodeSandbox.

[![Edit RCV - Variable Size](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-variable-size-8vu3u?fontsize=14&hidenavigation=1&theme=dark)

```js
import useVirtual from "react-cool-virtual";

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 1000,
    itemSize: (idx) => (idx % 2 ? 100 : 50),
  });

  return (
    <div
      style={{ width: "300px", height: "300px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, size }) => (
          <div key={index} style={{ height: `${size}px` }}>
            ⭐️ {index}
          </div>
        ))}
      </div>
    </div>
  );
};
```

### Dynamic Size

This example demonstrates how to create a dynamic (unknown) size row. For column or grid, please refer to CodeSandbox.

[![Edit RCV - Dynamic Size](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-dynamic-size-0wurg?fontsize=14&hidenavigation=1&theme=dark)

```js
import useVirtual from "react-cool-virtual";

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 1000,
    itemSize: 75, // The unmeasured item sizes will refer to this value (default = 50)
  });

  return (
    <div
      style={{ width: "300px", height: "300px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, measureRef }) => (
          // Use the `measureRef` to measure the item size
          <div key={index} ref={measureRef}>
            {/* Some data... */}
          </div>
        ))}
      </div>
    </div>
  );
};
```

> 💡 The scrollbar is jumping (or unexpected position)? It's because the total size of the items is gradually corrected along with an item that has been measured. You can tweak the `itemSize` to reduce the phenomenon.

### Real-time Resize

This example demonstrates how to create a real-time resize row (e.g. accordion, collapse, etc.). For column or grid, please refer to CodeSandbox.

[![Edit RCV - Real-time Resize](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-real-time-resize-fixvr?fontsize=14&hidenavigation=1&theme=dark)

```js
import { useState, forwardRef } from "react";
import useVirtual from "react-cool-virtual";

const AccordionItem = forwardRef(({ children, height, ...rest }, ref) => {
  const [h, setH] = useState(height);

  return (
    <div
      {...rest}
      style={{ height: `${h}px` }}
      ref={ref}
      onClick={() => setH((prevH) => (prevH === 50 ? 100 : 50))}
    >
      {children}
    </div>
  );
});

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 50,
  });

  return (
    <div
      style={{ width: "300px", height: "300px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, size, measureRef }) => (
          // Use the `measureRef` to measure the item size
          <AccordionItem key={index} height={size} ref={measureRef}>
            👋🏻 Click Me
          </AccordionItem>
        ))}
      </div>
    </div>
  );
};
```

### Responsive Web Design (RWD)

This example demonstrates how to create a list with RWD to provide a better UX for the user.

[![Edit RCV - RWD](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-rwd-x6lvc?fontsize=14&hidenavigation=1&theme=dark)

```js
import useVirtual from "react-cool-virtual";

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 1000,
    // Use the outer's width (2nd parameter) to adjust the item's size
    itemSize: (_, width) => (width > 400 ? 50 : 100),
    // The event will be triggered on outer's size changes
    onResize: (size) => console.log("Outer's size: ", size),
  });

  return (
    <div
      style={{ width: "100%", height: "400px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {/* We can also access the outer's width here */}
        {items.map(({ index, size, width }) => (
          <div key={index} style={{ height: `${size}px` }}>
            ⭐️ {index} ({width})
          </div>
        ))}
      </div>
    </div>
  );
};
```

> 💡 If the item size is specified through the function of `itemSize`, please ensure there's no the [measureRef](#items) on the item element. Otherwise, the hook will use the measured (cached) size for the item. When working with RWD, we can only use either of the two.

### Sticky Headers

This example demonstrates how to make sticky headers with React Cool Virtual.

[![Edit RCV - Sticky Headers](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-sticky-headers-pm79x?fontsize=14&hidenavigation=1&theme=dark)

```js
import useVirtual from "react-cool-virtual";

const List = () => {
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: 1000,
    itemSize: 75,
    stickyIndices: [0, 10, 20, 30, 40, 50], // The values must be provided in ascending order
  });

  return (
    <div
      style={{ width: "300px", height: "300px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, size, isSticky }) => {
          let style = { height: `${size}px` };
          // Use the `isSticky` property to style the sticky item, that's it ✨
          style = isSticky ? { ...style, position: "sticky", top: "0" } : style;

          return (
            <div key={someData[index].id} style={style}>
              {someData[index].content}
            </div>
          );
        })}
      </div>
    </div>
  );
};
```

> 💡 The scrollbar disappears when using Chrome in Mac? If you encounter [this issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1033712), you can add `will-change:transform` to the outer element to workaround this problem.

### Scroll to Offset / Items

You can imperatively scroll to offset or items as follows:

[![Edit RCV - Scroll-to Methods](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-scroll-to-methods-cs9x9?fontsize=14&hidenavigation=1&theme=dark)

```js
const { scrollTo, scrollToItem } = useVirtual();

const scrollToOffset = () => {
  // Scrolls to 500px
  scrollTo(500, () => {
    // 🤙🏼 Do whatever you want through the callback
  });
};

const scrollToItem = () => {
  // Scrolls to the 500th item
  scrollToItem(500, () => {
    // 🤙🏼 Do whatever you want through the callback
  });

  // We can control the alignment of the item with the `align` option
  // Acceptable values are: "auto" (default) | "start" | "center" | "end"
  // Using "auto" will scroll the item into the view at the start or end, depending on which is closer
  scrollToItem({ index: 10, align: "auto" });
};
```

### Smooth Scrolling

React Cool Virtual provides the smooth scrolling feature out of the box, all you need to do is turn the `smooth` option on.

[![Edit RCV - Smooth Scrolling](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-smooth-scrolling-noy4d?fontsize=14&hidenavigation=1&theme=dark)

```js
const { scrollTo, scrollToItem } = useVirtual();

// Smoothly scroll to 500px
const scrollToOffset = () => scrollTo({ offset: 500, smooth: true });

// Smoothly scroll to the 500th item
const scrollToItem = () => scrollToItem({ index: 10, smooth: true });
```

> 💡 When working with [dynamic size](#dynamic-size), the scroll position will be automatically corrected along with the items are measured. To optimize it, we can provide an estimated item size to the [itemSize](#itemsize) option.

The default easing effect is [easeInOutSine](https://easings.net/#easeInOutSine), and the duration is `100ms <= distance * 0.075 <= 500ms`. You can easily customize your own effect as follows:

```js
const { scrollTo } = useVirtual({
  // For 500 milliseconds
  scrollDuration: 500,
  // Or whatever duration you want based on the scroll distance
  scrollDuration: (distance) => distance * 0.05,
  // Using "easeInOutBack" effect (default = easeInOutSine), see: https://easings.net/#easeInOutSine
  scrollEasingFunction: (t) => {
    const c1 = 1.70158;
    const c2 = c1 * 1.525;

    return t < 0.5
      ? (Math.pow(2 * t, 2) * ((c2 + 1) * 2 * t - c2)) / 2
      : (Math.pow(2 * t - 2, 2) * ((c2 + 1) * (t * 2 - 2) + c2) + 2) / 2;
  },
});

const scrollToOffset = () => scrollTo({ offset: 500, smooth: true });
```

> 💡 For more cool easing effects, please [check it out](https://easings.net).

### Infinite Scroll

It's possible to make a complicated infinite scroll logic simple by just using a hook, no kidding! Let's see how possible 🤔.

[![Edit RCV - Infinite Scroll](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-infinite-scroll-3y351?fontsize=14&hidenavigation=1&theme=dark)

#### Working with [Skeleton Screens](https://uxdesign.cc/what-you-should-know-about-skeleton-screens-a820c45a571a)

```js
import { useState } from "react";
import useVirtual from "react-cool-virtual";
import axios from "axios";

const TOTAL_COMMENTS = 500;
const BATCH_COMMENTS = 5;
const isItemLoadedArr = [];

const loadData = async ({ loadIndex }, setComments) => {
  // Set the state of a batch items as `true`
  // to avoid the callback from being invoked repeatedly
  isItemLoadedArr[loadIndex] = true;

  try {
    const { data: comments } = await axios(`/comments?postId=${loadIndex + 1}`);

    setComments((prevComments) => {
      const nextComments = [...prevComments];

      comments.forEach((comment) => {
        nextComments[comment.id - 1] = comment;
      });

      return nextComments;
    });
  } catch (err) {
    // If there's an error set the state back to `false`
    isItemLoadedArr[loadIndex] = false;
    // Then try again
    loadData({ loadIndex }, setComments);
  }
};

const List = () => {
  const [comments, setComments] = useState([]);
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: TOTAL_COMMENTS,
    // Estimated item size (with padding)
    itemSize: 122,
    // The number of items that you want to load/or pre-load, it will trigger the `loadMore` callback
    // when the user scrolls within every items, e.g. 1 - 5, 6 - 10, and so on (default = 15)
    loadMoreCount: BATCH_COMMENTS,
    // Provide the loaded state of a batch items to the callback for telling the hook
    // whether the `loadMore` should be triggered or not
    isItemLoaded: (loadIndex) => isItemLoadedArr[loadIndex],
    // We can fetch the data through the callback, it's invoked when more items need to be loaded
    loadMore: (e) => loadData(e, setComments),
  });

  return (
    <div
      style={{ width: "300px", height: "500px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, measureRef }) => (
          <div
            key={comments[index]?.id || `fb-${index}`}
            style={{ padding: "16px", minHeight: "122px" }}
            ref={measureRef} // Used to measure the unknown item size
          >
            {comments[index]?.body || "⏳ Loading..."}
          </div>
        ))}
      </div>
    </div>
  );
};
```

#### Working with A Loading Indicator

```js
import { Fragment, useState } from "react";
import useVirtual from "react-cool-virtual";
import axios from "axios";

const TOTAL_COMMENTS = 500;
const BATCH_COMMENTS = 5;
const isItemLoadedArr = [];
// We only have 50 (500 / 5) batches of items, so set the 51th (index = 50) batch as `true`
// to avoid the `loadMore` callback from being invoked, yep it's a trick 😉
isItemLoadedArr[50] = true;

const loadData = async ({ loadIndex }, setComments) => {
  isItemLoadedArr[loadIndex] = true;

  try {
    const { data: comments } = await axios(`/comments?postId=${loadIndex + 1}`);

    setComments((prevComments) => [...prevComments, ...comments]);
  } catch (err) {
    isItemLoadedArr[loadIndex] = false;
    loadData({ loadIndex }, setComments);
  }
};

const Loading = () => <div>⏳ Loading...</div>;

const List = () => {
  const [comments, setComments] = useState([]);
  const { outerRef, innerRef, items } = useVirtual({
    itemCount: comments.length, // Provide the number of comments
    loadMoreCount: BATCH_COMMENTS,
    isItemLoaded: (loadIndex) => isItemLoadedArr[loadIndex],
    loadMore: (e) => loadData(e, setComments),
  });

  return (
    <div
      style={{ width: "300px", height: "500px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.length ? (
          items.map(({ index, measureRef }) => {
            const showLoading =
              index === comments.length - 1 && comments.length < TOTAL_COMMENTS;

            return (
              <Fragment key={comments[index].id}>
                <div ref={measureRef}>{comments[index].body}</div>
                {showLoading && <Loading />}
              </Fragment>
            );
          })
        ) : (
          <Loading />
        )}
      </div>
    </div>
  );
};
```

### Pre-pending Items

This example demonstrates how to pre-pend items and maintain scroll position for the user.

[![Edit RCV - Prepend Items](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-prepend-items-ui06h?fontsize=14&hidenavigation=1&theme=dark)

```js
import { useEffect, useLayoutEffect, useState } from "react";

import useVirtual from "react-cool-virtual";
import axios from "axios";

const TOTAL_COMMENTS = 500;
const BATCH_COMMENTS = 5;
let shouldFetchData = true;
let postId = 100;

const fetchData = async (postId, setComments) => {
  try {
    const { data: comments } = await axios(`/comments?postId=${postId}`);

    // Pre-pend new items
    setComments((prevComments) => [...comments, ...prevComments]);
  } catch (err) {
    // Try again
    fetchData(postId, setComments);
  }
};

const List = () => {
  const [comments, setComments] = useState([]);
  const { outerRef, innerRef, items, startItem } = useVirtual({
    // Provide the number of comments
    itemCount: comments.length,
    onScroll: ({ scrollForward, scrollOffset }) => {
      // Tweak the threshold of data fetching that you want
      if (!scrollForward && scrollOffset < 50 && shouldFetchData) {
        fetchData(--postId, setComments);
        shouldFetchData = false;
      }
    },
  });

  useEffect(() => fetchData(postId, setComments), []);

  // Execute the `startItem` through `useLayoutEffect` before the browser to paint
  // See https://reactjs.org/docs/hooks-reference.html#uselayouteffect to learn more
  useLayoutEffect(() => {
    // After the list updated, maintain the previous scroll position for the user
    startItem(BATCH_COMMENTS, () => {
      // After the scroll position updated, re-allow data fetching
      if (comments.length < TOTAL_COMMENTS) shouldFetchData = true;
    });
  }, [comments.length, startItem]);

  return (
    <div
      style={{ width: "300px", height: "500px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.length ? (
          items.map(({ index, measureRef }) => (
            // Used to measure the unknown item size
            <div key={comments[index].id} ref={measureRef}>
              {comments[index].body}
            </div>
          ))
        ) : (
          <div className="item">⏳ Loading...</div>
        )}
      </div>
    </div>
  );
};
```

### Filtering Items

When working with filtering items, we can reset the scroll position when the `itemCount` is changed by enabling the [resetScroll](#resetscroll) option.

[![Edit RCV - Filter Items](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-filter-items-clno9?fontsize=14&hidenavigation=1&theme=dark)

```js
import { useState } from "react";
import useVirtual from "react-cool-virtual";

const List = () => {
  const [itemCount, setItemCount] = useState(100);
  const { outerRef, innerRef, items } = useVirtual({
    itemCount,
    // Resets the scroll position when the `itemCount` is changed (default = false)
    resetScroll: true,
  });

  return (
    <div
      style={{ width: "300px", height: "300px", overflow: "auto" }}
      ref={outerRef}
    >
      <div ref={innerRef}>
        {items.map(({ index, size }) => (
          <div key={index} style={{ height: `${size}px` }}>
            ⭐️ {index}
          </div>
        ))}
      </div>
    </div>
  );
};
```

### Sticking to Bottom

This example demonstrates the scenario of sticking/unsticking the scroll position to the bottom for a chatroom.

[![Edit RCV - Stick to Bottom](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/rcv-stick-to-bottom-cxsgw?fontsize=14&hidenavigation=1&theme=dark)

```js
import { useState, useEffect } from "react";
import useVirtual from "react-cool-virtual";
import axios from "axios";

const TOTAL_MESSAGES = 200;
let isScrolling = false; // Used to prevent UX conflict
let id = 0;

const loadData = async (id, setMessages) => {
  try {
    const { data: messages } = await axios(`/messages/${id}`);


CHANGELOG.md
# react-cool-virtual

## 0.7.0

### Minor Changes

- [#376](https://github.com/wellyshen/react-cool-virtual/pull/376) [`744953b`](https://github.com/wellyshen/react-cool-virtual/commit/744953bcafd376c78c8621ebd6421b87b62a3aef) Thanks [@wellyshen](https://github.com/wellyshen)! - feat: use `startItem` instead `startItemIndex` for pre-pending items

## 0.6.0

### Minor Changes

- [#368](https://github.com/wellyshen/react-cool-virtual/pull/368) [`d1f15c9`](https://github.com/wellyshen/react-cool-virtual/commit/d1f15c96c82449c77e9119971819821aec853046) Thanks [@wellyshen](https://github.com/wellyshen)! - feat: add `startItemIndex` method for pre-pending items

## 0.5.8

### Patch Changes

- [`9ee5ae5`](https://github.com/wellyshen/react-cool-virtual/commit/9ee5ae528244cfdd4c7903274680a6a62f18e0b4) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: the `e.userScroll` of `onScroll` and `loadMore` should be `false` when using `scrollTo` or `scrollToItem` with smooth mode

## 0.5.7

### Patch Changes

- [#353](https://github.com/wellyshen/react-cool-virtual/pull/353) [`70e0581`](https://github.com/wellyshen/react-cool-virtual/commit/70e0581940b763302c1d6bb630e2993d04391e1c) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: wrong item length when using `ssrItemCount` with an array type value

## 0.5.6

### Patch Changes

- [#172](https://github.com/wellyshen/react-cool-virtual/pull/172) [`1970c62`](https://github.com/wellyshen/react-cool-virtual/commit/1970c625dc28412c21a27ab98745210e825a61f6) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: scrolling flicking in React v18

## 0.5.5

### Patch Changes

- [#285](https://github.com/wellyshen/react-cool-virtual/pull/285) [`41878e3`](https://github.com/wellyshen/react-cool-virtual/commit/41878e3e18a87ccd26d6ea91e99a462ddcfff486) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: correct the event properties of `loadMore` when scrolling backward

## 0.5.4

### Patch Changes

- [#256](https://github.com/wellyshen/react-cool-virtual/pull/256) [`281a31e`](https://github.com/wellyshen/react-cool-virtual/commit/281a31ef500360c41829098db892f873633fbedd) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: filter items not updated correctly

## 0.5.3

### Patch Changes

- [#254](https://github.com/wellyshen/react-cool-virtual/pull/254) [`c453cad`](https://github.com/wellyshen/react-cool-virtual/commit/c453cad0c82e7e40baf741dd8f0676360e378041) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: app crash when dynamic change data (from short to long)

## 0.5.2

### Patch Changes

- [#252](https://github.com/wellyshen/react-cool-virtual/pull/252) [`9467444`](https://github.com/wellyshen/react-cool-virtual/commit/9467444b2eb558d3e23d74a5b61cedd471b06232) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: filter data with dynamic size causes empty row/column

## 0.5.1

### Patch Changes

- [#245](https://github.com/wellyshen/react-cool-virtual/pull/245) [`f31007e`](https://github.com/wellyshen/react-cool-virtual/commit/f31007e71f0228f3c82ec35ea7de9a60d0a51850) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: app crash due to filtering (from more to less)

## 0.5.0

### Patch Changes

- [#231](https://github.com/wellyshen/react-cool-virtual/pull/231) [`b705ab3`](https://github.com/wellyshen/react-cool-virtual/commit/b705ab30a11cecebe9ed7bd8f2579d12f6fd9381) Thanks [@nikitapilgrim](https://github.com/nikitapilgrim)! - fix: `stickyIndices` can be dynamic

## 0.4.6

### Patch Changes

- [#225](https://github.com/wellyshen/react-cool-virtual/pull/225) [`60d4e78`](https://github.com/wellyshen/react-cool-virtual/commit/60d4e786788d3b9c590abdf89a76166e92e84dc2) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: app crashes when scrolling with drag bar

## 0.4.4

### Patch Changes

- [#183](https://github.com/wellyshen/react-cool-virtual/pull/183) [`737996f`](https://github.com/wellyshen/react-cool-virtual/commit/737996ff65ba8f58c7380931b6b350c0c0b6c044) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: correct the initial `e.userScroll` of `loadMore` callback from `true` to `false`

## 0.4.3

### Patch Changes

- [#181](https://github.com/wellyshen/react-cool-virtual/pull/181) [`44769e5`](https://github.com/wellyshen/react-cool-virtual/commit/44769e575acda5fcbdbe4c82cd0f5348d14434eb) Thanks [@wellyshen](https://github.com/wellyshen)! - refactor: simplify generic type setup

* [#179](https://github.com/wellyshen/react-cool-virtual/pull/179) [`41ab2a7`](https://github.com/wellyshen/react-cool-virtual/commit/41ab2a7253cb33762dc389f5661c52ebde6ca62c) Thanks [@wellyshen](https://github.com/wellyshen)! - fix: `scrollToItem` width dynamic size causes infinitely loop

## 0.4.2

### Patch Changes

- [#173](https://github.com/wellyshen/react-cool-virtual/pull/173) [`70d04ac`](https://github.com/wellyshen/react-cool-virtual/commit/70d04ac5ebab7a686e58eca670679178b2e93a17) Thanks [@wellyshen](https://github.com/wellyshen)! - perf(useVirtual): mutate inner element styles only when items are updated

## 0.4.1

### Patch Changes

- [#155](https://github.com/wellyshen/react-cool-virtual/pull/155) [`72f2a6c`](https://github.com/wellyshen/react-cool-virtual/commit/72f2a6cc64e6cbe585d37e9e104f9179b4d55c27) Thanks [@wellyshen](https://github.com/wellyshen)! - perf(useVirtual): remove unnecessary internal `useDebounce` for reducing bundle size

* [#154](https://github.com/wellyshen/react-cool-virtual/pull/154) [`abc5880`](https://github.com/wellyshen/react-cool-virtual/commit/abc5880d83cb0b3e83b0a56b2ed688abf950f00a) Thanks [@wellyshen](https://github.com/wellyshen)! - fix(useVirtual): do not trigger `onScroll` and `loadMore` by internally scroll-to method

## 0.4.0

### Minor Changes


index.js
/* eslint-disable jsx-a11y/accessible-emoji */

import { useEffect, useState } from "react";
import { render } from "react-dom";

import useVirtual from "react-cool-virtual";
import axios from "axios";
import "./styles.scss";

const TOTAL_COMMENTS = 500;
const BATCH_COMMENTS = 5;
let shouldFetchData = true;
let postId = 100;

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

const fetchData = async (postId, setComments) => {
  try {
    // Simulating a slow network
    await sleep(500);

    const { data: comments } = await axios(
      `https://jsonplaceholder.typicode.com/comments?postId=${postId}`
    );

    // Pre-pend new items
    setComments((prevComments) => [...comments, ...prevComments]);
  } catch (err) {
    // Try again
    fetchData(postId, setComments);
  }
};

const App = () => {
  const [comments, setComments] = useState([]);
  const { outerRef, innerRef, items, scrollToItem } = useVirtual({
    // Provide the number of comments
    itemCount: comments.length,
    // When working with dynamic size, we can reduce scroll jumping
    // by providing an estimated item size
    itemSize: 180,
    onScroll: ({ scrollOffset }) => {
      // Tweak the threshold of data fetching that you want
      if (scrollOffset < 50 && shouldFetchData) {
        fetchData(--postId, setComments);
        shouldFetchData = false;
      }
    }
  });

  useEffect(() => {
    const fetchInitData = async () => {
      await fetchData(postId, setComments);
      await fetchData(--postId, setComments);
    };

    fetchInitData();
  }, []);

  useEffect(() => {
    scrollToItem({ index: BATCH_COMMENTS, align: "start" }, () => {
      // After the scroll position updated, re-allow data fetching
      if (comments.length < TOTAL_COMMENTS) shouldFetchData = true;
    });
  }, [comments.length, scrollToItem]);

  return (
    <div className="app">
      <p>
        💡 CodeSandbox might be <b>caching the old version</b> of React Cool
        Virtual. You can <b>manually specify to the latest version</b> to see
        the examples.
      </p>
      <br />
      <div
        className="outer"
        style={{ width: "300px", height: "500px", overflow: "auto" }}
        ref={outerRef}
      >
        <div ref={innerRef}>
          {items.length ? (
            items.map(({ index, measureRef }) => (
              <div
                key={comments[index].id}
                className="item"
                ref={measureRef} // Used to measure the unknown item size
              >
                {comments[index].id}. {comments[index].body}
              </div>
            ))
          ) : (
            <div className="item">⏳ Loading...</div>
          )}
        </div>
      </div>
      <br />
      <br />
    </div>
  );
};

render(<App />, document.getElementById("root"));
